piker/ui/_search.py

```diff
@@ -0,0 +1,961 @@
+        self,
+    ) -> tuple[QModelIndex, QStandardItem]:
+
+        model = self.model()
+        isections = model.rowCount()
+
+        # much thanks to following code to figure out breadth-first
+        # traversing from the root node:
+        # https://stackoverflow.com/a/33126689
+        for i in range(isections):
+            idx = model.index(i, 0, QModelIndex())
+            item = model.itemFromIndex(idx)
+            yield idx, item
+
+    def find_section(
+        self,
+        section: str,
+
+    ) -> Optional[QModelIndex]:
+        '''Find the *first* depth = 1 section matching ``section`` in
+        the tree and return its index.
+
+        '''
+        for idx, item in self.iter_d1():
+            if item.text() == section:
+                return idx
+        else:
+            # caller must expect his
+            return None
+
+    def clear_section(
+        self,
+        section: str,
+        status_field: str = None,
+
+    ) -> None:
+        '''Clear all result-rows from under the depth = 1 section.
+
+        '''
+        idx = self.find_section(section)
+        model = self.model()
+
+        if idx is not None:
+            if model.hasChildren(idx):
+                rows = model.rowCount(idx)
+                # print(f'removing {rows} from {section}')
+                assert model.removeRows(0, rows, parent=idx)
+
+            # remove section as well ?
+            # model.removeRow(i, QModelIndex())
+
+            if status_field is not None:
+                model.setItem(idx.row(), 1, QStandardItem(status_field))
+
+            else:
+                model.setItem(idx.row(), 1, QStandardItem())
+
+            self.resize()
+
+            return idx
+        else:
+            return None
+
+    def set_section_entries(
+        self,
+        section: str,
+        values: Sequence[str],
+        clear_all: bool = False,
+
+    ) -> None:
+        '''Set result-rows for depth = 1 tree section ``section``.
+
+        '''
+        model = self.model()
+        if clear_all:
+            # XXX: rewrite the model from scratch if caller requests it
+            model.clear()
+
+        model.setHorizontalHeaderLabels(self.labels)
+
+        section_idx = self.clear_section(section)
+
+        # if we can't find a section start adding to the root
+        if section_idx is None:
+            root = model.invisibleRootItem()
+            section_item = QStandardItem(section)
+            blank = QStandardItem('')
+            root.appendRow([section_item, blank])
+
+        else:
+            section_item = model.itemFromIndex(section_idx)
+
+        # values just needs to be sequence-like
+        for i, s in enumerate(values):
+
+            ix = QStandardItem(str(i))
+            item = QStandardItem(s)
+
+            # Add the item to the model
+            section_item.appendRow([ix, item])
+
+        self.expandAll()
+
+        # TODO: figure out if we can avoid this line in a better way
```

Comment:
This bit here running on every section results insert is what's causing the selection to jump around when re-searches take place while typing. it's super annoying and basically we want to just select whatever the *last* selection was if possible to avoid this.

Reply:
made another comment on how we could try figuring out the "abstract row,colum, parent" position once the user starts navigating and even if entries are swapped out underneath we just jump to the same "relative selection position" when possible.

Again can be worked out in #186.